pkg/controllers/providers/instancetype/capacity/suite_test.go

```diff
@@ -174,4 +209,65 @@ var _ = Describe("CapacityCache", func() {
 		mem.Sub(resource.MustParse(fmt.Sprintf("%dMi", int64(math.Ceil(float64(mem.Value())*options.FromContext(ctx).VMMemoryOverheadPercent/1024/1024)))))
 		Expect(i.Capacity.Memory().Value()).To(Equal(mem.Value()), "Expected capacity to match VMMemoryOverheadPercent calculation")
 	})
+
+	It("should properly update discovered capacity when matching AMI is not the first in the list", func() {
+		// Update nodeClass AMIs for this test
+		nodeClass.Status.AMIs = []v1.AMI{standardAMI}
+		ExpectApplied(ctx, env.Client, nodeClass)
+
+		// Get available instance types from the test environment
+		availableInstanceTypes, err := awsEnv.InstanceTypesProvider.List(ctx, nodeClass)
+		Expect(err).To(BeNil())
+		Expect(availableInstanceTypes).ToNot(BeEmpty(), "No instance types available in test environment")
+
+		// Choose the first instance type for testing
+		testInstanceType := availableInstanceTypes[0]
+		instanceTypeName := testInstanceType.Name
+
+		fmt.Printf("Initial memory capacity for %s: %v\n", testInstanceType.Name, testInstanceType.Capacity.Memory().Value())
+
+		// Create a test node with the discovered instance type
+		memoryCapacity := resource.MustParse("4Gi")
+
+		testNodeClassNvidiaFirst := test.EC2NodeClass()
+		testNodeClassNvidiaFirst.Status.AMIs = []v1.AMI{nvidiaAMI, standardAMI}
+		ExpectApplied(ctx, env.Client, testNodeClassNvidiaFirst)
+
+		testNodeClaim := coretest.NodeClaim()
+		testNodeClaim.Spec.NodeClassRef = &karpv1.NodeClassReference{
+			Group: object.GVK(testNodeClassNvidiaFirst).Group,
+			Kind:  object.GVK(testNodeClassNvidiaFirst).Kind,
+			Name:  testNodeClassNvidiaFirst.Name,
+		}
+		testNodeClaim.Status.Capacity = corev1.ResourceList{
+			corev1.ResourceMemory: memoryCapacity,
+		}
+		testNodeClaim.Labels[corev1.LabelInstanceTypeStable] = instanceTypeName
+		testNodeClaim.Labels[corev1.LabelArchStable] = karpv1.ArchitectureAmd64
+		testNodeClaim.Status.ImageID = "ami-standard-test"
+		ExpectApplied(ctx, env.Client, testNodeClaim)
+
+		testNode := coretest.NodeClaimLinkedNode(testNodeClaim)
+		ExpectApplied(ctx, env.Client, testNode)
+
+		ExpectObjectReconciled(ctx, env.Client, controller, testNode)
+
+		// Verify that the cache was updated by getting the instance types and checking the memory capacity
+		instanceTypesAfterUpdateReversed, err := awsEnv.InstanceTypesProvider.List(ctx, testNodeClassNvidiaFirst)
+		Expect(err).To(BeNil())
+
+		// Find our instance type and verify its memory capacity was updated
+		found := false
+		for _, it := range instanceTypesAfterUpdateReversed {
+			if it.Name == instanceTypeName {
+				found = true
+				fmt.Printf("Discovered memory capacity for %s: %v\n", it.Name, it.Capacity.Memory().Value())
```

Comment:
Same for this print, we generally don't leave prints in the tests and just rely on the expectations to communicate issues.